Add sequence splitting option (#161)
triton-inference-server/dali_backend: docs/examples/video_decoding/README.md

# Decoding video in DALI backend

## Splitting sequences into subsequences
Often the video model we want to run inference on expects sequences of some fixed lentght. In such case we may want to split the decoded video
into a batch of sub-sequences of that given length. We can achieve that utilizing the `fn.reshape` operator from DALI together with a model parameter
`"split_along_outer_axis"` set in configuration file. Lets see an example of using this combination.

```python
FRAMES_PER_SEQUENCE = 5
OUT_WIDTH = 300
OUT_HEIGHT = 300

@autoserialize
@dali.pipeline_def(batch_size=256, num_threads=4, device_id=0,
                   output_dtype=dali.types.UINT8, output_ndim=5)
def pipeline():
  vid = fn.external_source(device='cpu', name='INPUT', ndim=1, dtype=dali.types.UINT8)
  seq = fn.experimental.decoders.video(vid, device='mixed')
  seq = fn.resize(seq, resize_x=OUT_WIDTH, resize_y=OUT_HEIGHT)

>   seq = fn.pad(seq, axis_names='F', align=FRAMES_PER_SEQUENCE)

Author: Resize is kind of necessary, because we need to know frame size to perform the reshape.
It wouldn't be needed if we knew that all of our files have some known resolution

  return fn.reshape(seq, shape=[-1, FRAMES_PER_SEQUENCE, OUT_HEIGHT, OUT_WIDTH, 3], name='OUTPUT')
```

> You can see that we added two operators at the end of the pipeline - `pad` and `reshape`. Padding samples is needed because we want to split them into sub-sequneces of `FRAMES_PER_SEQUENCE` frames each - that means that we need to ensure that the total number of frames in a sample is divisible by `FRAMES_PER_SEQUENCE`. The result of this operator is adding frames filled with zeroes to the end of the sample to align to that number.

Reviewer: There's a lot of `that` word in the second sentence. How about rephrasing?
```suggestion
You can see that we added two operators at the end of the pipeline - `pad` and `reshape`. Since we want to split the sequences into sub-sequneces of `FRAMES_PER_SEQUENCE` frames each, we must ensure that the total number of frames in every sample is divisible by `FRAMES_PER_SEQUENCE`. The Pad operator adds frames filled with zeroes to the end of every sample to align to that number.
```

> After padding the samples, we use the `reshape` operator to add another dimension to the data, so layout-wise the transformation could be represented as `FHWC` -> `NFHWC`, where `N` dimension represents index of a sub-sequence. An example flow could be described as follows: lets say a sample after resizing has a shape of `(13, 300, 300, 3)` - it's a sequence of 13 frames. Assuming the `FRAMES_PER_SEQUENCE` is 5, the sample is then padded to shape `(15, 300, 300, 3)`. Then, the reshape, without modifying any data, changes it to `(3, 5, 300, 300, 3)` - 3 sub-sequences with 5 frames each.

Reviewer: How about expressing this transformation using the example shapes in a pipeline. I mean something like:
```python
def pipeline():
  vid = fn.external_source(device='cpu', name='INPUT', ndim=1, dtype=dali.types.UINT8)  # (6, 2859274)
  seq = fn.experimental.decoders.video(vid, device='mixed')  # (6, 300, 300, 3)
  seq = fn.pad(seq, axis_names='F', align=FRAMES_PER_SEQUENCE)  # (10, 300, 300, 3)
  return fn.reshape(seq, shape=[-1, FRAMES_PER_SEQUENCE, OUT_HEIGHT, OUT_WIDTH, 3], name='OUTPUT')  # (2, 5, 300, 300, 3)
```
This might be attached to the example above, or maybe specified in some pseudo code, or a chain of operations? Like `input(6, 2859274) -> decode(6, 300, 300, 3)` etc...

The point here would be to guide the user what are the shape changes across the pipeline.

Reviewer: Alternatively, you could specify a concrete "problem" at the top of this document and follow this example throughout. Something for example:
```
Let's say we have a 3 video files, which have different number of frames: 20, 13 and 18. The model we want to use always takes a sequence of 4 frames as an input
```
That's just a suggestion, it may make this description better understandable.

Author: I did it as python snippet interleaved with comments describing the changes of sample shape

> Reshaping data in a described way deosn't solve our problem completely but we can combine it with an additional model parameter that will split the output tensors and treat each sub-sequence as a separate sample in the batch.

Reviewer: > Reshaping data in a described way deosn't solve our problem completely

Could you describe, why such reshaping does not solve the problem? I mean, I know why, but I believe it would be good to put it here.

Author: done, I think

> Reshaping data in a described way deosn't solve our problem completely but we can combine it with an additional model parameter that will split the output tensors and treat each sub-sequence as a separate sample in the batch.

Reviewer: ```suggestion
Reshaping data in a described way doesn't solve our problem completely, but we can combine it with an additional model parameter that will split the output tensors and treat each sub-sequence as a separate sample in the batch.
```

> The parameter is called `split_along_outer_axis` and can be set in the configuration file as follows:

Reviewer: ```suggestion
The parameter is called `split_along_outer_axis` and can be set in the model configuration file as follows:
```

```pbtxt
parameters [
  {
    key: "split_along_outer_axis",

>     value: { string_value: "OUTPUT" }

Reviewer: ```suggestion
    value: { string_value: "OUTPUT_1:OUTPUT_2" }
```

Reviewer: This brings me a thought, would `{ string_value: "OUTPUT_1:" }` work? What would happen if user provides this syntax?

Author: I wanted the config file to match the pipeline, so I included only single output. 

Yup, "OUTPUT_1:" would work

  }
]
```

> The value of the parameter (specified as `string_value`) is an output name (or list of names seperated with a colon) that should be transformed. We can see the effect of this parameter on the following example: lets assume the DALI pipeline produced a batch of two samples with the shapes {(1, 5, 300, 300, 3), (2, 5, 300, 300, 3)}, so a sample with 1 sub-sequence and a sample with 2 sub-sequences. The batch produced by our model will then have shape {(5, 300, 300, 3), (5, 300, 300, 3), (5, 300, 300, 3)} - a batch of 3 sequences.

Reviewer: ```suggestion
The value of the parameter (specified as `string_value`) is the name of the output (or list of names separated with a colon) that should be transformed. We can see the effect of this parameter on the following example: let's assume the DALI pipeline produced a batch of two samples with the shapes `{(1, 5, 300, 300, 3), (2, 5, 300, 300, 3)}`, so the first sample with 1 sub-sequence and the second sample with 2 sub-sequences. The batch produced by our model will then have shape `{(5, 300, 300, 3), (5, 300, 300, 3), (5, 300, 300, 3)}` - a batch of 3 sequences.
```

> As you can see, the said parameter helps us also handle video files of different lenghts. Outputting such sequences by the DALI pipeline is not feasible in Triton because batches in Triton need to have uniform shape (every sample in the batch should have the same shape). Using the `reshape` - `"split_along_outer_axis"` combination can transform unevenly shaped sequences into a batch of evenly shaped sub-sequences.

Reviewer: ```suggestion
As you can see, the said parameter helps us also handle video files of different lenghts. Outputting such sequences by the DALI pipeline is not feasible in Triton, because batches in Triton need to have a uniform shape (every sample in the batch should have the same shape). Using the `reshape` + `"split_along_outer_axis"` combination can transform unevenly shaped sequences into a batch of evenly shaped sub-sequences.
```
